wgpu-hal/src/lib.rs

```diff
@@ -522,6 +535,60 @@ pub trait Adapter: WasmNotSendSync {
     unsafe fn get_presentation_timestamp(&self) -> wgt::PresentationTimestamp;
 }
 
+/// A connection to a GPU and a pool of resources to use with it.
+///
+/// A `wgpu-hal` `Device` represents an open connection to specific graphics
+/// processor, controlled via the backend [`Device::A`]. This corresponds to a
+/// backend object like a Vulkan logical device ([`vkDevice`]) or a Direct3D 12
+/// virtual adapter ([`ID3D12Device`]).
+///
+/// A `Device`'s life cycle is generally:
+///
+/// - Obtain a `Device` and its associated [`Queue`] by calling
+///   [`Adapter::open`].
+///
+///   Alternatively, the backend-specific types that implement [`Adapter`] often
+///   have methods for creating a `wgpu-hal` `Device` from a platform-specific
+///   handle. For example, [`vulkan::Adapter::device_from_raw`] can create a
+///   [`vulkan::Device`] from an [`ash::Device`].
+///
+/// - Create resources to use on the device by calling methods like
+///   [`Device::create_texture`] or [`Device::create_shader_module`].
+///
+/// - Call [`Device::create_command_encoder`] to obtain a [`CommandEncoder`],
+///   which you can use to build [`CommandBuffer`]s holding commands to be
+///   executed on the GPU.
+///
+/// - Call [`Queue::submit`] on the `Device`'s associated [`Queue`] to submit
+///   [`CommandBuffer`]s for execution on the GPU. If needed, call
+///   [`Device::wait`] to wait for them to finish execution.
+///
+/// - Free resources with methods like [`Device::destroy_texture`] or
+///   [`Device::destroy_shader_module`].
+///
+/// - Shut down the device by calling [`Device::exit`].
+///
+/// [`vkDevice`]: https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VkDevice
+/// [`ID3D12Device`]: https://learn.microsoft.com/en-us/windows/win32/api/d3d12/nn-d3d12-id3d12device
+/// [`vulkan::Adapter::device_from_raw`]: vulkan/struct.Adapter.html#method.device_from_raw
+/// [`vulkan::Device`]: vulkan/struct.Device.html
+/// [`ash::Device`]: https://docs.rs/ash/latest/ash/struct.Device.html
+/// [`CommandBuffer`]: Api::CommandBuffer
+///
+/// # Safety
+///
+/// As usual for `wgpu-hal`, [validation] is the caller's responsibility. Here
```

Comment:
nitpick: The phrasing "As usual…" doesn't seem appropriate for the target audience, which is to introduce to or remind of to safety considerations when using `wgpu-hal`. I feel like it might be better to be more concrete, like "As with other `wgpu-hal` APIs…".

```suggestion
/// As with other `wgpu-hal` APIs, [validation] is the caller's responsibility. Here
```